AbstractSQLCompiler.ts

```diff
@@ -370,10 +378,150 @@ const compileSchema = (abstractSqlModel: AbstractSqlModel, engine: Engines, ifNo
 	}
 }
 
+const generateSplit = <T>(src:Array<T>, dst: Array<T>, matchFn: MatchFn<T>): Split<T> => {
+	let modified: Array<Pair<T>> = []
+	return _.reduce(src, (acc, value) => {
+		const match = matchFn(acc.inserted, value)
+		if (_.isUndefined(match)) {
+			return acc
+		} else {
+			acc.inserted = _.without(acc.inserted, match)
+			acc.deleted = _.without(acc.deleted, value)
+			acc.modified.push( { src: value, dst: match } )
+			return acc
+		} 
+	}
+	, { inserted: dst, deleted: src, modified: modified } )
+}
+//  genDiff expects src and dst arrays formatted as described in
+//  genSplit, and three functions to be called in the case of
+//  insertion, deletion or modification.
+//  The match function describes how to find the corresponding T inside a T[].
+//  The result of the match function will be either the element that should be matched to 
+//  the argument or undefined if no such element is found.
+const generateDiff = <T>(insFn: ToSQLFn<T>, delFn: ToSQLFn<T>, modFn: ToSQLFn<Pair<T>>, matchFn: MatchFn<T>, src: Array<T>, dst: Array<T>) => {
+	const split = generateSplit(src, dst, matchFn)
+
+	const diff = _.map(split.modified, modFn)
+	.concat(_.map(split.deleted, delFn))
+	.concat(_.map(split.inserted, insFn))
+	
+	return _.reject(diff, _.isUndefined)
+}
+
+const diffFields = (src: AbstractSqlField[], dst: AbstractSqlField[], mappings: ResourceMap<string>, engine: Engines, ifNotExists: boolean) => {
+	let ifNotExistsStr: string
+	let ifExistsStr: string
+
+	if (ifNotExists) {
+		ifNotExistsStr = 'IF NOT EXISTS '
+		ifExistsStr = 'IF EXISTS ' 
+	} else {
+		ifNotExistsStr = ''
+		ifExistsStr = ''
+	}
+
+	const matchFn: MatchFn<AbstractSqlField> = (fieldArray, field) => {
+		let match = _.find(fieldArray, { fieldName: field.fieldName })
+		if (_.isUndefined(match)) {
+			if (_.isString(mappings[field.fieldName])) {
+				return _.find(fieldArray, { fieldName: mappings[field.fieldName] } )
+			}
+		} else {
+			return match
+		}
+	}
+
+	const insFn: ToSQLFn<AbstractSqlField> = (field) => {
+		// Columns are inserted as NULL for the time being, this is because we have no sensible way to automatically assign a default value
+		return 'ADD COLUMN ' + ifNotExistsStr + '"' + field.fieldName + '" ' + dataTypeGen(engine, field) + ';'
+		
+	}
+
+	const delFn: ToSQLFn<AbstractSqlField> = (field) => {
+		return 'DROP COLUMN ' + ifExistsStr + '"' + field.fieldName + '";'
+	}
+	
+	const modFn: ToSQLFn<Pair<AbstractSqlField>> = ({src, dst}) => {
+		if (_.isEqual(src, dst)) {
+			return
+		}
+		if (_.isEqual(_.omit(src, ['fieldName', 'references']), _.omit(dst, ['fieldName', 'references']))) {
+			return 'RENAME COLUMN "' + src.fieldName + '" TO "' + dst.fieldName + '";'
+		}
+		throw Error(`Can not migrate pre-existing field ${src.fieldName} to ${dst.fieldName}`)
```

Comment:
Could you add the data type to the message, as you can get here with the field name being the same but data type different, and the current error message would be confusing in that case